addOns/websocket/src/main/java/org/zaproxy/zap/extension/websocket/pscan/WebSocketPassiveScannerManager.java

```diff
@@ -0,0 +1,252 @@
+/*
+ * Zed Attack Proxy (ZAP) and its related class files.
+ *
+ * ZAP is an HTTP/HTTPS proxy for assessing web application security.
+ *
+ * Copyright 2018 The ZAP Development Team
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.zaproxy.zap.extension.websocket.pscan;
+
+import java.util.Iterator;
+import java.util.concurrent.CopyOnWriteArraySet;
+import org.apache.log4j.Logger;
+import org.zaproxy.zap.extension.websocket.WebSocketObserver;
+import org.zaproxy.zap.extension.websocket.WebSocketSenderListener;
+import org.zaproxy.zap.extension.websocket.alerts.AlertManager;
+import org.zaproxy.zap.extension.websocket.db.TableWebSocket;
+
+/**
+ * Manages all stuff related with the WebSocket Passive Scanning. Manager is able to open a
+ * background thread {@link WebSocketPassiveScanThread} so as to run passive scans. In addition, all
+ * {@link WebSocketPassiveScanner} implementations should be registered and enabled/disabled by this
+ * class. Manager keeps the WebSocket Scanners into thread safe list. Finally, informs every Passive
+ * Scanners (should implement {@link WebSocketSenderListener}) about messages with method{@link
+ * WebSocketPassiveScannerManager#getWebSocketScannerObserver()}
+ */
+public class WebSocketPassiveScannerManager {
+
+    private static final Logger LOGGER = Logger.getLogger(WebSocketPassiveScannerManager.class);
+
+    /** The background thread where the passive scans are running */
+    private WebSocketPassiveScanThread passiveScanThread;
+
+    /** {@code false} if passive scans are disabled */
+    private boolean isThreadActive = false;
+
+    /** Used to raise Alert Messages */
+    private AlertManager alertManager;
+
+    /** List stores passive scanners */
+    private CopyOnWriteArraySet<WebSocketPassiveScanner> passiveScannersSet;
```

Comment:
This can be a set of `WebSocketPassiveScannerDecorator` (internally it's just handled the decorator).

Reply:
However externally we have to return `Iterator<WebSocketPassiveScanner>`. Do you think the following code is reasonable?

```
 /** @return an iterator for all WebSocket Passive Scanners */
protected Iterator<WebSocketPassiveScanner> getIterator() {
    return getPassiveScannersSet().stream()
            .map(x -> x.getWebSocketPassiveScanner())
            .iterator();
}

```

Reply:
Right, but that's an internal method as well (not public) so fine to change to `WebSocketPassiveScannerDecorator`?

Reply:
My fault, I used this method as example. Mostly I am wondering about `getEnabledIterator()` which is use by `WebSocketPassiveScanThread`

Reply:
I think it's fine in that case too, in fact we can let `WebSocketPassiveScanThread` do the enabled check (and just have one `getIterator()`).

Reply:
Fixed in af11c8f